website/src/routes/guides/(migration)/migrate-from-zod/index.mdx

```diff
@@ -1,7 +1,101 @@
 ---
 title: Migrate from Zod
+description: Im based on Zod, so migrating should not require much work, because the naming and behaviour of methods is very similar.
 ---
 
+import { Link } from '@builder.io/qwik-city';
+
 # Migrate from Zod
 
-> The content of this page is not yet ready. Check back in a few days.
+I am heavily based on Zod, you could say it was the inspiration for my own creation. That's why a migration should not require much work, because the naming and behaviour of methods is very similar.
+
+## Setup
+
+First you need to install me. See <Link href="/guides/installation/">Installation instructions</Link>.
+
+## Schema Migration
+
+After this setup you will have to change every **Zod Schema**,
+but you don't have to change all of them just yet.
+
+> **This is a Process, so take your time!** *You can always take a break, and I will still be here, waiting for you.*
+
+Theres usually two ways to write these Schemas with Zod:
+
+### The `z` Object
+
+> Zod makes heavy use of it's `z` object export, which contains every schema type and validation function. This is one of Zod's disadvantages, because modern bundlers can't remove unused functions. Learn more <Link href="/guides/comparison/">here</Link>.
```

Comment:
This might be slightly confusing since the primary reason why Zod cannot support efficient tree shaking is not because it's using `z` object export. Improving this should be a low hanging fruits for Zod since they can just export symbols separately and enforce people to use `import * as z from 'zod'`. The primary reason is that they are using classes and most of features are implemented as methods, which cannot be tree shakable. They even have some inlined logic, which is even harder to tree shake as shown in [this screenshot](https://twitter.com/naruaway/status/1687006117430525952) (a tweet in Japanese)

(BTW finding these was the main reason why I got interested in Valibot)

Reply:
Why are they even using classes? I'm currently writing a custom translation library and I tried using classes but for every usecase objects just worked much better.

BTW, we could use something like the `addIssueToContext` function for removing the Exception performance penalty.
We also have to return some `INVALID` value, like Zod, to signal validation failure. I think this will make the code more complex and "ugly" but that should be ok for the performance benefit. Maybe there's an elegant solution?

Reply:
> Why are they even using classes?

I have no idea! But probably partially because to make it easier to implement "method chain". Or maybe just coding style preference. I personally always avoid classes though 😅 

> BTW, we could use something like the addIssueToContext function

Yes, I guess we can discuss this in https://github.com/fabian-hiller/valibot/issues/48 ! Now I am confident that Valibot can be faster than Zod except for failure case (https://github.com/fabian-hiller/valibot/issues/73#issuecomment-1669292338)